phat/contracts/ink_price_feed/lib.rs

```diff
@@ -0,0 +1,678 @@
+            };
+
+            let mut client = connect(config)?;
+
+            // Get a request if presents
+            let request: PriceRequestMessage = client
+                .pop()
+                .log_err("answer_price: failed to read queue")?
+                .ok_or(Error::NoRequestInQueue)?;
+
+            let response = Self::handle_request(&request)?;
+            // Attach an action to the tx by:
+            client.action(Action::Reply(response.encode()));
+
+            maybe_submit_tx(client, &self.attest_key, config.sender_key.as_ref())
+        }
+
+        fn handle_request(request: &PriceRequestMessage) -> Result<PriceResponseMessage> {
+            let trading_pair_id = request.trading_pair_id;
+            let token0 = request.token0.as_str();
+            let token1 = request.token1.as_str();
+
+            info!("Request received: ({trading_pair_id}, {token0}, {token1})");
+            // Get the price and respond as a rollup action.
+            match Self::fetch_coingecko_price(token0, token1) {
+                Ok(price) => {
+                    // Respond
+                    info!("Price: {price}");
+                    let response = PriceResponseMessage {
+                        resp_type: TYPE_RESPONSE,
+                        trading_pair_id,
+                        price: Some(price),
+                        err_no: None,
+                    };
+                    Ok(response)
+                }
+                // Error when fetching the price. Could be
+                Err(Error::FailedToDecode) => {
+                    warn!("Fail to decode the price");
+                    let response = PriceResponseMessage {
+                        resp_type: TYPE_ERROR,
+                        trading_pair_id,
+                        price: None,
+                        err_no: Some(0),
+                    };
+                    Ok(response)
+                }
+                Err(e) => Err(e),
+            }
+        }
+
+        /// Returns BadOrigin error if the caller is not the owner
+        fn ensure_owner(&self) -> Result<()> {
+            if self.env().caller() == self.owner {
+                Ok(())
+            } else {
+                Err(Error::BadOrigin)
+            }
+        }
+
+        /// Returns the config reference or raise the error `NotConfigured`
+        fn ensure_configured(&self) -> Result<&Config> {
+            self.config.as_ref().ok_or(Error::NotConfigured)
+        }
+    }
+
+    fn connect(config: &Config) -> Result<InkRollupClient> {
+        let result = InkRollupClient::new(
+            &config.rpc,
+            config.pallet_id,
+            config.call_id,
+            &config.contract_id,
+        )
+        .log_err("failed to create rollup client");
+
+        match result {
+            Ok(client) => Ok(client),
+            Err(e) => {
+                error!("Error : {:?}", e);
+                Err(Error::FailedToCreateClient)
+            }
+        }
```

Comment:
`.log_err` will print the error to the logger. So you can simply write like:

```suggestion
        InkRollupClient::new(
            &config.rpc,
            config.pallet_id,
            config.call_id,
            &config.contract_id,
        )
        .log_err("failed to create rollup client")
        .or(Err(Error::FailedToCreateClient))
```

Reply:
I think the root cause will not be displayed anymore

Reply:
Yes, it will be displayed. `log_err(msg)` will prints `${msg}: ${err}`.